core/src/entry_verification_service/worker.rs

```diff
@@ -0,0 +1,524 @@
+        io::Error,
+        ops::RangeInclusive,
+        result::Result,
+        sync::Arc,
+        thread,
+        time::{Duration, Instant},
+    },
+};
+
+mod commands;
+
+pub(super) enum Command {
+    /// Add new data set for the specified slot.
+    AddDataSet {
+        slot: Slot,
+        shard_indices: RangeInclusive<u32>,
+        last_in_slot: bool,
+        entries: Arc<Vec<Entry>>,
+    },
+    /// Indicates that someone is waiting for a validation result for this slot.  This is just a
+    /// hint to start validation ASAP.
+    WaitingFor {
+        slot: Slot,
+        data_set: DataSetFirstShred,
+    },
+    /// Remove state for slots before the specified one.
+    Cleanup { earlier_than: Slot },
+}
+
+pub(super) fn start(
+    blockstore: Arc<Blockstore>,
+    results: Arc<SharedResults>,
+) -> Result<crossbeam_channel::Sender<Command>, Error> {
+    let (commands_sender, commands_receiver) = crossbeam_channel::unbounded();
+
+    let mut state = State {
+        commands: commands_receiver,
+        blockstore,
+        processing: HashMap::new(),
+        waiting_for_parent_hash: HashMap::new(),
+        client_waiting_for: None,
+        results,
+    };
+
+    thread::Builder::new()
+        .name("solVerifyEntries".to_string())
+        .spawn(move || state.run())?;
+
+    Ok(commands_sender)
+}
+
+/// Verification status for a single slot.
+#[derive(PartialEq, Eq, Debug, Clone)]
+enum SlotStatus {
+    /// Some entries int he slot are not fully verified yet.
+    Processing {
+        /// Parent slot index.  It is used to get a hash for the PoH verification of the very first
+        /// [`Entry`] in this slot.
+        parent_slot: Slot,
+
+        /// Index of the first data set, starting from the beginning of the slot, that has not been
+        /// fully verified yet, with no gaps.  When this value exceeds `last_data_set`, slot is
+        /// considered verified, and the state changes to `Verified`.
+        verified_up_to: DataSetFirstShred,
+
+        /// Index of the last data set in this slot.  We only get to know it when we receive a data
+        /// set holding the very last entry.  At that point we are going to set this field.
+        last_data_set: Option<DataSetFirstShred>,
+
+        /// Holds optimistic status for an individual data set.
+        ///
+        /// If any of the pending verification fails, the whole slot is marked as [`Failed`].
+        ///
+        /// Entries are delivered in chunks, called "data sets".  Each data set is contained in
+        /// a consecutive range of shreds, that were used to deliver those entries.
+        ///
+        /// This map only contains entries at or after `verified_up_to`.
+        data_sets: HashMap<DataSetFirstShred, DataSetOptimisticStatus>,
+    },
+
+    /// All entries in the slot have been successfully verified.  We still need to remember the very
+    /// last entry hash, as it is needed for verification of any slots that use this one as a
```

Comment:
*needed for verification of any children of this slot

Reply:
Applied.